packages/discovery-react-components/src/components/DocumentPreview/components/PdfViewerHighlight/utils/textBoxMapping/CellProvider.ts

```diff
@@ -0,0 +1,96 @@
+import { isSideBySideOnLine } from '../common/bboxUtils';
+import { TextLayoutCell, TextLayoutCellBase } from '../textLayout/types';
+
+export class CellProvider {
+  private readonly skippedCells: TextLayoutCellBase[] = [];
+  private cells: TextLayoutCellBase[]; // make sure to handle this as immutable array
+  private cursor: number = 0;
+
+  constructor(cells: TextLayoutCellBase[]) {
+    this.cells = [...cells];
+  }
+
+  hasNext() {
+    while (this.cursor < this.cells.length) {
+      const cell = this.cells[this.cursor];
+      if (cell.text.trim().length !== 0) {
+        break;
+      }
+      this.skip();
+    }
+    return this.cursor < this.cells.length;
+  }
+
+  /** get cells on a line */
+  private getNextCells = (() => {
+    let lastCells: TextLayoutCellBase[] | null = null;
+    let lastCursor: number | null = null;
+    let lastResult: TextLayoutCellBase[] | null = null;
+
+    return () => {
+      if (lastResult && lastCells === this.cells && lastCursor === this.cursor) {
+        return lastResult;
+      }
+
+      const result: TextLayoutCellBase[] = [];
+      let lastCell: TextLayoutCell | null = null;
+      for (let i = this.cursor; i < this.cells.length; i += 1) {
+        const currentBox = this.cells[i];
+        // maybe we need to break this loop by big box change
+        const { cell: baseCurrentCell } = currentBox.getNormalized();
+        if (lastCell && !isSideBySideOnLine(lastCell.bbox, baseCurrentCell.bbox)) {
+          break;
+        }
+        result.push(currentBox);
+        lastCell = baseCurrentCell;
+      }
+      lastCells = this.cells;
+      lastCursor = this.cursor;
+      lastResult = result;
+
+      return result;
+    };
+  })();
```

Comment:
I'm confused by this setup: the function returns a function but it is immediately invoked? Why not just have a normal return?

Reply:
As the method is expensive and can be called multiple times, I implemented a cache. The setup is to keep variables related to the cache in local scope. I agree it's not common and will rewrite it.